diff --git a/config.sub b/config.sub
index e4c4e09..6dc3666 100644
--- a/config.sub
+++ b/config.sub
@@ -1768,7 +1768,7 @@ case $os in
 	     | onefs* | tirtos* | phoenix* | fuchsia* | redox* | bme* \
 	     | midnightbsd* | amdhsa* | unleashed* | emscripten* | wasi* \
 	     | nsk* | powerunix* | genode* | zvmoe* | qnx* | emx* | zephyr* \
-	     | fiwix* | mlibc* | cos* | mbr* | ironclad* )
+	     | fiwix* | mlibc* | cos* | mbr* | ironclad* | ohos* )
 		;;
 	# This one is extra strict with allowed versions
 	sco3.2v2 | sco3.2v[4-9]* | sco5v6*)
@@ -1866,6 +1866,8 @@ case $kernel-$os-$obj in
 		;;
 	*-eabi*- | *-gnueabi*-)
 		;;
+	*-ohos*-)
+		;;
 	none--*)
 		# None (no kernel, i.e. freestanding / bare metal),
 		# can be paired with an machine code file format
diff --git a/configure b/configure
index 488d93f..55ff910 100644
--- a/configure
+++ b/configure
@@ -17042,6 +17042,10 @@ else
 	INSTALL_OPT_S=
 fi
 
+if test "${with_strip_program+set}" = set; then
+  INSTALL_OPT_S="$INSTALL_OPT_S --strip-program=$with_strip_program"
+fi
+
 : "${INSTALL:=install}"
 echo "$as_me:17046: checking if install accepts -p option" >&5
 echo $ECHO_N "checking if install accepts -p option... $ECHO_C" >&6
