diff -Nur a/misc/gen-pkgconfig.in b/misc/gen-pkgconfig.in
--- a/misc/gen-pkgconfig.in	2020-03-08 09:57:33.000000000 +0800
+++ b/misc/gen-pkgconfig.in	2020-04-14 14:12:37.824000000 +0800
@@ -83,7 +83,7 @@
 fi
 
 lib_flags=
-for opt in -L$libdir @LDFLAGS@ @EXTRA_LDFLAGS@ @LIBS@
+for opt in -L$libdir @LIBS@
 do
 	case $opt in
 	-l*) # LIBS is handled specially below
diff -Nur a/misc/ncurses-config.in b/misc/ncurses-config.in
--- a/misc/ncurses-config.in	2020-02-03 07:34:34.000000000 +0800
+++ b/misc/ncurses-config.in	2020-04-14 14:15:53.844000000 +0800
@@ -41,7 +41,6 @@
 
 bindir="@bindir@"
 includedir="@includedir@"
-libdir="@libdir@"
 datarootdir="@datarootdir@"
 datadir="@datadir@"
 mandir="@mandir@"
@@ -101,7 +100,7 @@
 # There is no portable way to find the list of standard library directories. 
 # Require a POSIX shell anyway, to keep this simple.
 lib_flags=
-for opt in -L$libdir @LDFLAGS@ @EXTRA_LDFLAGS@ $LIBS
+for opt in $LIBS
 do
 	case $opt in
 	-specs*) # ignore linker specs-files which were used to build library
@@ -113,9 +112,6 @@
 	-L*)
 		[ -d ${opt##-L} ] || continue
 		case ${opt##-L} in
-		@LD_SEARCHPATH@) # skip standard libdir
-			continue
-			;;
 		*)
 			found=no
 			for check in $lib_flags
@@ -235,7 +231,6 @@
 		echo $INCS
 		;;
 	--libdir)
-		echo "${libdir}"
 		;;
 	--mandir)
 		echo "${mandir}"
